exercises/python/encryption-context-complete/src/document_bucket/api.py

```diff
@@ -0,0 +1,106 @@
+                    PointerQuery)
+
+
+class DocumentBucketOperations:
+    def __init__(self, bucket, table, master_key_provider: KMSMasterKeyProvider):
+        self.bucket = bucket
+        self.table = table
+        self.master_key_provider: KMSMasterKeyProvider = master_key_provider
+
+    def _write_pointer(self, item: PointerItem):
+        self.table.put_item(Item=item.to_item())
+
+    def _write_object(self, data: bytes, item: PointerItem):
+        s3object = self.bucket.put_object(
+            Body=data, Key=item.partition_key, Metadata=item.context
+        )
+        return s3object
+
+    def _get_object(self, item: PointerItem) -> bytes:
+        s3object = self.bucket.Object(item.partition_key).get()
+        return s3object["Body"].read()
+
+    def _populate_key_records(self, pointer: PointerItem) -> Set[ContextItem]:
+        context_items: Set[ContextItem] = pointer.context_items()
+        for context_item in context_items:
+            self.table.put_item(Item=context_item.to_item())
+        return context_items
+
+    def _query_for_context_key(self, query: ContextQuery) -> Set[PointerItem]:
+        result = self.table.query(KeyConditionExpression=query.expression())
+        pointers: Set[PointerItem] = set()
+        for ddb_context_item in result["Items"]:
+            context_item = ContextItem.from_item(ddb_context_item)
+            pointer_query = PointerQuery.from_context_item(context_item)
+            pointer_items = self.table.query(
+                KeyConditionExpression=pointer_query.expression()
+            )["Items"]
+            if len(pointer_items) != 1:
+                raise ValueError(
+                    "Pointer ID not unique! Expected 1 result, got {}".format(
+                        len(pointer_items)
+                    )
+                )
+            pointer = PointerItem.from_item(pointer_items[0])
+            pointers.add(pointer)
+        return pointers
+
+    def _scan_table(self) -> Set[PointerItem]:
+        result = self.table.scan(PointerItem.filter_for())
+        pointers = set()
+        for ddb_item in result["Items"]:
+            pointer = PointerItem.from_item(ddb_item)
+            pointers.add(pointer)
+        return pointers
+
+    def list(self) -> Set[PointerItem]:
+        return self._scan_table()
+
+    def retrieve(
+        self,
+        pointer_key: str,
+        expected_context_keys: Set[str] = set(),
+        expected_context: Dict[str, str] = {},
+    ) -> DocumentBundle:
+        item = PointerItem.from_key_and_context(pointer_key, expected_context)
+        encrypted_data = self._get_object(item)
+        (plaintext, header) = aws_encryption_sdk.decrypt(
+            source=encrypted_data, key_provider=self.master_key_provider
+        )
+        if expected_context_keys not in header.encryption_context.keys():
+            raise AssertionError(
+                "Encryption context assertion failed! Expected: {} but got {}".format(
```

Comment:
What version of Python are we targeting? Since Python 3.6+ is what's supported today, can we use f-strings here instead? https://realpython.com/python-f-strings/

Reply:
Version: 3.8+, baby. B)

F-strings: Oh heck yeah, will change. I just muscle memory'd.